File: Arrays/119/119.c
#include <stdio.h>
#include <stdlib.h>

int *getRow1(int rowIndex, int* returnSize){
    *returnSize = rowIndex + 1;
    int *row = (int *)malloc(sizeof(int) * (rowIndex + 1));
    if(rowIndex == 0){
        row[0] = 1;
    }else{
        int *prevRow = getRow1(rowIndex - 1, returnSize);
        for(int i = 1; i <= rowIndex; i++){
            row[i] += prevRow[i - 1] + prevRow[i];
        }
        row[0] = 1;
        free(prevRow);
    }
    return row;   
}

int *getRow(int rowIndex, int* returnSize){
    *returnSize = rowIndex + 1;
    int *row = (int *)malloc(sizeof(int) * (rowIndex + 1));
    row[0] = 1;
    long temp = 1;
    for(int i=1, up=rowIndex; i<=rowIndex; i++, up--){
        temp = temp * up / i;
        row[i] = temp;
    }
    return row;
}

int main(){
    int x = 6;
    int *row = getRow(5, &x);
    for(int i = 0; i < 6; i++){
        printf("%d ", row[i]);
    }
    printf("\n");
}

File: Arrays/119/119.py
from typing import List

# O(n^2) solution
class Solution1:
    def getRow(self, rowIndex: int) -> List[int]:
        row = [1]
        if rowIndex == 0:
            return row
        
        prevRow = Solution1.getRow(self, rowIndex - 1)
        for i in range(1, len(prevRow)):
            curr = prevRow[i - 1] + prevRow[i]
            row.append(curr)
        row.append(1)
        return row

# O(n) solution
class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        row = [1] * (rowIndex + 1)
        up = rowIndex
        down = 1
        for i in range(1, rowIndex):
            row[i] = row[i-1] * up // down
            up -= 1
            down += 1
        return row
    
sol = Solution()
sol1 = Solution1()
print(sol.getRow(5))
print(sol1.getRow(5))